File: LeetCode/Challenges/2021/January_LeetCoding_Challenge/java/20.valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        HashMap<Character, Character> openingBracket = new HashMap<>();
        openingBracket.put(')', '(');
        openingBracket.put('}', '{');
        openingBracket.put(']', '[');
        
        for (int i = 0; i < s.length(); ++i) {
            switch(s.charAt(i)) {
                case '(':
                case '{':
                case '[':
                    st.push(s.charAt(i));
                    break;

                default:
                    if (st.empty()) return false;
                    if (st.peek() != openingBracket.get(s.charAt(i))) return false;
                    st.pop();
                    break;
            }
        }
        return st.empty();
    }
}

File: LeetCode/Challenges/2021/January_LeetCoding_Challenge/cpp/20.valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        unordered_map<char, char> openingBracket;
        openingBracket[')'] = '(';
        openingBracket['}'] = '{';
        openingBracket[']'] = '[';
        
        for (char ch: s) {
            switch(ch) {
                case '(':
                case '{':
                case '[':
                    st.push(ch);
                    break;

                default:
                    if (st.empty()) return false;
                    if (st.top() != openingBracket[ch]) return false;
                    st.pop();
                    break;
            }
        }
        return st.empty();
    }
};